Allow installing modules from a vcs subdirectory in non-editable mode

sbidoul opened the pull request:
Fixes #3217 

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/pypa/pip/3466)
<!-- Reviewable:end -->

xavfernandez commented on pip/index.py:
```diff
@@ -932,6 +932,15 @@ def egg_fragment(self):
         r'(sha1|sha224|sha384|sha256|sha512|md5)=([a-f0-9]+)'
     )
 
+    _subdirectory_fragment_re = re.compile(r'[#&]subdirectory=([^&]*)')
```
Maybe move this after `def hash` method to keep the regex followed by its method pattern.
Maybe also update `_egg_fragment_re = re.compile(r'#egg=([^&]*)')` to `_egg_fragment_re = re.compile(r'[#&]egg=([^&]*)')`

sbidoul replied:
Yes, of course, fixed.

sbidoul commented:
I factored out the code in a setup_py_dir property, and adapted the wheel command too.

xavfernandez commented:
Great job ! It's much cleaner :+1: 

This also means you can remove `_build_editable_options` and all references to `editable_options`.

And maybe add a few tests in https://github.com/pypa/pip/blob/61f75c40adc06fdb5aa21d2a1e3c0101b83077a9/tests/unit/test_index.py#L41 for `subdirectory_fragment`

sbidoul commented:
@xavfernandez I love adding features while removing code at the same time :smile: 

I removed editable_options, added the test for link fragments, and added subdirectory support for pip download.

Now I don't understand why travis fails on pypy, previous run was ok. Must be a transient error?

xavfernandez commented on pip/req/req_install.py:
```diff
@@ -1201,21 +1165,10 @@ def parse_editable(editable_req, default_vcs=None):
+            '#egg=Package' % editable_req
         )
-    if not options or 'egg' not in options:
-        req = _build_req_from_url(editable_req)
```
Hmm it looks like this should not go away (or it should be indicated as backward incompatible).

sbidoul replied:
@xavfernandez ok, I restored the call. All green now.

xavfernandez replied:
:+1:

xavfernandez commented:
Yes the pypy failure was due to some flakiness on PyPI apparently.

xavfernandez commented:
Thanks a lot !
